internal/p10k.zsh

```diff
@@ -4280,24 +4281,56 @@ function instant_prompt_chezmoi_shell() {
 # Virtualenv: current working virtualenv
 # More information on virtualenv (Python):
 # https://virtualenv.pypa.io/en/latest/
+
+_virtualenv_VIRTUAL_ENV() {
+  local n
+  # VIRTUAL_ENV is the on-disk path, VIRTUAL_ENV_PROMPT is `(name_of_virtualenv) `.
+  # Use custom name if it was set (python -m venv -p "foo" .venv)
+  if [[ $VIRTUAL_ENV_PROMPT == '('?*') ' && $VIRTUAL_ENV_PROMPT != "($n) " ]]; then
+    n=$VIRTUAL_ENV_PROMPT[2,-3]
+  # Use parent directory name if virtualenv name is generic (eg .venv)
+  elif [[ $v == $~_POWERLEVEL9K_VIRTUALENV_GENERIC_NAMES ]]; then
+    n=${VIRTUAL_ENV:h:t}
+  # Otherwise use the virtualenv name as-is
+  else
+    n=${VIRTUAL_ENV:t}
+  fi
+  echo "$n"
+}
+_virtualenv_poetry() {
+  local idx=$1
+  local dir=${_p9k__parent_dirs[idx]}
+  local pyproject="$dir/pyproject.toml"
+  _p9k_cached_cmd 0 '' poetry -C "$dir" version
+  # Return the first word only, eg the value of pyproject.toml's `poetry.name`
+  echo "${_p9k__ret%% *}"
+}
 prompt_virtualenv() {
   local msg=''
   if (( _POWERLEVEL9K_VIRTUALENV_SHOW_PYTHON_VERSION )) && _p9k_python_version; then
     msg="${_p9k__ret//\%/%%} "
   fi
-  local v=${VIRTUAL_ENV:t}
-  if [[ $VIRTUAL_ENV_PROMPT == '('?*') ' && $VIRTUAL_ENV_PROMPT != "($v) " ]]; then
-    v=$VIRTUAL_ENV_PROMPT[2,-3]
-  elif [[ $v == $~_POWERLEVEL9K_VIRTUALENV_GENERIC_NAMES ]]; then
-    v=${VIRTUAL_ENV:h:t}
+  # Determine virtualenv name with a few strategies
+  local n=''
+  if [[ -n $VIRTUAL_ENV ]]; then
+    n=$(_virtualenv_VIRTUAL_ENV)
+  else
+    if [[ $POWERLEVEL9K_VIRTUALENV_ENABLE_POETRY == true ]]; then
```

Comment:
I added this feature flag to help with testing. Reminder to self: remove before merging. I can't think of a reason to disable this functionality.

Reply:
done